osquery/filesystem/file_compression.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#ifdef WIN32
+#define _WIN32_DCOM
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#endif
+
+// This define is required for Windows static linking of libarchive
+#define LIBARCHIVE_STATIC
+#include <archive.h>
+#include <archive_entry.h>
+#include <zstd.h>
+
+#include <osquery/flags.h>
+#include <osquery/system.h>
+
+#include "osquery/carver/carver.h"
+#include "osquery/filesystem/fileops.h"
+
+namespace osquery {
+
+DECLARE_uint32(carver_block_size);
+
+Status compress(const boost::filesystem::path& in,
+                const boost::filesystem::path& out) {
+  PlatformFile inFile(in.string(), PF_OPEN_EXISTING | PF_READ);
+  PlatformFile outFile(out.string(), PF_CREATE_NEW | PF_WRITE);
+  ZSTD_CStream* const cstream = ZSTD_createCStream();
+  if (cstream == nullptr) {
+    return Status(1, "Couldn't create compression stream");
+  }
+
+  size_t const initResult = ZSTD_initCStream(cstream, 1);
+  if (ZSTD_isError(initResult)) {
+    return Status(1, "Couldn't initialize compression stream");
+  }
+
+  size_t const buffInSize = ZSTD_CStreamInSize();
+  size_t const buffOutSize = ZSTD_CStreamOutSize();
+  std::vector<void*> buffIn(buffInSize);
+  std::vector<void*> buffOut(buffOutSize);
+  auto read = buffInSize;
+  auto toRead = buffInSize;
+
+  while (true) {
```

Comment:
This would just be instrumenting a max file size for carver which is probably not something we want to do here. In future we can use carver to compress large regions of memory as well  so I would say if we want to do this, let's do it before this loop.